gdal/ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp

```diff
@@ -806,6 +809,79 @@ OGRErr OGRGeoPackageTableLayer::ReadTableDefinition()
         }
     }
 
+#if !defined(__GNUC__) || defined(__clang__) || __GNUC__ >= 5
+
+    // Unique fields detection
+    char* pszTableDefinitionSQL = sqlite3_mprintf("select sql from sqlite_master where type='table' and name='%q'", m_pszTableName);
+    err = SQLQuery(poDb, pszTableDefinitionSQL, &oResultTable);
+    sqlite3_free(pszTableDefinitionSQL);
+
+    if ( err != OGRERR_NONE || oResultTable.nRowCount == 0 )
+    {
+        if( oResultTable.pszErrMsg != nullptr )
+            CPLError( CE_Failure, CPLE_AppDefined, "%s", oResultTable.pszErrMsg );
+        else
+            CPLError( CE_Failure, CPLE_AppDefined, "Cannot find table %s", m_pszTableName );
+
+        SQLResultFree(&oResultTable);
+        return OGRERR_FAILURE;
+    }
+
+    std::set<std::string> uniqueFields;
+    // Match identifiers with " or ` or no delimiter (and no spaces).
+    static const std::regex sFieldIdentifierRe { R"raw(\s*(["`]([^"`]+)["`])|(([^\s]+)\s).*)raw" };
+    std::string tableDefinition { SQLResultGetValue(&oResultTable, 0, 0) };
+    tableDefinition = tableDefinition.substr(tableDefinition.find('('), tableDefinition.rfind(')') );
+    std::stringstream tableDefinitionStream { tableDefinition };
+    std::smatch uniqueFieldMatch;
+    while (tableDefinitionStream.good()) {
+      std::string fieldStr;
+      getline( tableDefinitionStream, fieldStr, ',' );
+      if( fieldStr.find( "UNIQUE" ) != std::string::npos )
```

Comment:
perhaps test in case insensitive way ?

Reply:
I tried to use mixed case UniQue when creating the table and the index and it always got back upper case but I don't really know if it was my application (`sqlitemanager`) that did the trick, I'll test case insensitive.

Reply:
it is probably sqlite3 itself normalizing this as the below experiment seems to show:
```
$ sqlite3 foo.db
SQLite version 3.11.0 2016-02-15 17:29:24
Enter ".help" for usage hints.
sqlite> create table foo(bar integer);
sqlite> create unique index idx on foo(bar);
sqlite> 

$ sqlite3 foo.db
SQLite version 3.11.0 2016-02-15 17:29:24
Enter ".help" for usage hints.
sqlite> select * from sqlite_master;
table|foo|foo|2|CREATE TABLE foo(bar integer)
index|idx|foo|3|CREATE UNIQUE INDEX idx on foo(bar)

```

Reply:
ok, so no need to test case insensitive? Or better do it just in case different sqlite versions handle this differently?

Reply:
not here at least. but if doing "create table foo ( bar integer unique);", unique gets back at lowercase when reading sqlite_master

Reply:
Ok, I've implemented case insensitive already.